### PR TITLE
Add CMake support to build the libraries

### PR DESCRIPTION
**Pull request by gcuendet**
# Description

This PR adds the CMake support to build the libraries and the torchtrtc executable. It also generates a `torchtrtConfig.cmake`, in order to be able to consume these libraries with cmake targets.

The major advantage of the CMake support as introduced in this PR is the added ability to compile for windows as well, fixing the export of symbols, and generating both the `*.dll` and the `*.lib` libraries. To export the symbols, I reused the existing `TORCHTRT_API` macro, redefining it for windows, using CMake. The consumption of the dynamic libraries (both on linux and windows) is demonstrated by building the executable torchtrtc[.exe]. This has been tested locally with MSVC 2019 (19.29.30143.0), libtorch 1.11 (1.11.0+cu113), CUDA 11.6, cuDNN 8.2.1, and TensorRT 8.2.4.2.
A secondary advantage is the handling of dependencies. These can be installed anywhere (not necessarily globally on the system) and by any mean and compiling the lib does not require editing config file anymore, instead, a few CMake variables are enough to have the dependencies correctly found and used. (for cuDNN and TensorRT, most of the magic happens in the CMake finders in `cmake/Modules`).

This PR only adds partial CMake support to Torch-TensorRT (only for the libs and the executable torchtrtc). I started working on the unit tests as well and could consider working on adding the python support as well, if there is a clear interest. I still think that being able to compile the libraries (also on windows) has value in itself, and that's the reason for this PR.

I didn't add any documentation about how to compile using CMake, but that should probably be done. A bit of guidance on where and how this should be best done would be appreciated.

With this PR, I also would like to trigger the discussion around CMake and/or windows support (which, technically, are obviously totally independent). The kind of questions I have are:
- Is there a CI for this lib (I am assuming there is)? Could we imagine also compiling the lib with Cmake on the CI, to check that any new addition does not break it in the future? If not, how do you see the maintenance of the feature?
- Same question for windows: I am assuming the lib is not compiled on windows on the CI currently. Can that be added? Possibly with CMake?

Thanks for reviewing and looking forward to answer questions 😃 

## Type of change

- New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [ ] New and existing unit tests pass locally with my changes

Signed-off-by: Gabriel Cuendet <gabriel.cuendet@gmail.com>

### CODE REVIEW

File: core/partitioning/shape_analysis.cpp [hunk 1]
```diff
@@ -1,5 +1,6 @@
 #include "core/partitioning/shape_analysis.h"
 #include <ATen/ATen.h>
+#include <inttypes.h>
```
Comment (narendasan) [inline core/partitioning/shape_analysis.cpp:3]: Where does this file come from?

File: cpp/include/torch_tensorrt/macros.h [hunk 2]
```diff
@@ -11,8 +11,9 @@
 #define TORCHTRT_API __attribute__((__visibility__("default")))
 #define TORCHTRT_HIDDEN __attribute__((__visibility__("hidden")))
 #else
-#define TORCHTRT_API
-#define TORCHTRT_HIDDEN
+// This can be used also if defined(__GNUC__) but is generated by CMake
+// and will be missing when compiling with Bazel
+#include <torch_tensorrt_export.h>
```
Comment (narendasan) [inline cpp/include/torch_tensorrt/macros.h:16]: Can we #ifdef this based on CMake being the build system if it is not going to be generated by bazel?

File: core/conversion/converters/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,46 @@
+set(sub_lib_name "converters")
+
+target_sources(${lib_name} 
+    PRIVATE "${CMAKE_CURRENT_SOURCE_DIR}/NodeConverterRegistry.cpp"
+            "${CMAKE_CURRENT_SOURCE_DIR}/converter_util.cpp"
+            "${CMAKE_CURRENT_SOURCE_DIR}/Weights.cpp"
+            "${CMAKE_CURRENT_SOURCE_DIR}/impl/activation.cpp"
```
Comment (narendasan) [inline core/conversion/converters/CMakeLists.txt:7]: Could we just glob the impl directory here?

File: core/lowering/passes/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,33 @@
+target_sources(${lib_name} 
+    PRIVATE "${CMAKE_CURRENT_SOURCE_DIR}/convNd_to_convolution.cpp"
```
Comment (narendasan) [inline core/lowering/passes/CMakeLists.txt:2]: Also here can we just glob the full directory *.cpp?

File: cpp/include/torch_tensorrt/torch_tensorrt.h [hunk 5]
```diff
@@ -15,6 +15,8 @@
 #include <string>
 #include <vector>
 
+#include <torch_tensorrt/macros.h>
```
Comment (narendasan) [inline cpp/include/torch_tensorrt/torch_tensorrt.h:18]: Just style wise, we use quotes for all the project internal includes

**Comment by narendasan** [comment]
@gcuendet Can you sign your commits?

**Comment by narendasan** [comment]
> This PR only adds partial CMake support to Torch-TensorRT (only for the libs and the executable torchtrtc). I started working on the unit tests as well and could consider working on adding the python support as well, if there is a clear interest. I still think that being able to compile the libraries (also on windows) has value in itself, and that's the reason for this PR.

I think adding a `--use-cmake` flag to the `setup.py` to use the CMake system might be a good addition. I am not too sure about the unit tests as we are still working to get bazel working on Windows and I would rather keep the alternative build systems as light weight as possible. Guess it depends on how much extra work it is to add cmake tests.

> I didn't add any documentation about how to compile using CMake, but that should probably be done. A bit of guidance on where and how this should be best done would be appreciated.

`//docsrc/tutorials/installation.rst` would be the spot for this. You can make a CMake section or something.


>    Is there a CI for this lib (I am assuming there is)? Could we imagine also compiling the lib with Cmake on the CI, to check that any new addition does not break it in the future? If not, how do you see the maintenance of the feature?

 cc: @andi4191, I think we could just try to build it and just report if succeeds or not 

>    Same question for windows: I am assuming the lib is not compiled on windows on the CI currently. Can that be added? Possibly with CMake?

Windows CI is part of our testing plan down the road as support matures.

File: cpp/include/torch_tensorrt/macros.h (hunk 2 above)
Comment (gcuendet) [reply cpp/include/torch_tensorrt/macros.h]: Yes absolutely. Very good point. This is now done in [this commit](https://github.com/pytorch/TensorRT/pull/1058/commits/bc693d86f5c52958a5cb4b34262e0761a7ed2d4a).

File: core/partitioning/shape_analysis.cpp (hunk 1 above)
Comment (gcuendet) [reply core/partitioning/shape_analysis.cpp]: I think this is a header from the standard C library. But actually it's not the one that was needed, so it's been replaced by `#include <cstdint>` in [this commit](https://github.com/pytorch/TensorRT/pull/1058/commits/fc13eb2c36f59e4ef122f28bd4476092370403cd).

This was added to address the type problem described in issue #856 :
```
input_shapes.push_back(util::toVec(util::toDims(c10::List<long int>({1}))));
 // replaced by 
input_shapes.push_back(util::toVec(util::toDims(c10::List<long long>({1}))));
```

File: cpp/include/torch_tensorrt/torch_tensorrt.h (hunk 5 above)
Comment (gcuendet) [reply cpp/include/torch_tensorrt/torch_tensorrt.h]: This has been fixed in [this commit](https://github.com/pytorch/TensorRT/pull/1058/commits/c0bb7a8a18cfdfb84570891415c7fe66c604160b).

File: core/conversion/converters/CMakeLists.txt (hunk 3 above)
Comment (gcuendet) [reply core/conversion/converters/CMakeLists.txt:7]: Yes we could, however globbing source files in CMake is **not** a good practice, as documented in the [CMake documentation](https://cmake.org/cmake/help/latest/command/file.html?highlight=glob#file) itself:
> Note: We do not recommend using GLOB to collect a list of source files from your source tree. If no CMakeLists.txt file changes when a source is added or removed then the generated build system cannot know when to ask CMake to regenerate. The CONFIGURE_DEPENDS flag may not work reliably on all generators, or if a new generator is added in the future that cannot support it, projects using it will be stuck. Even if CONFIGURE_DEPENDS works reliably, there is still a cost to perform the check on every rebuild.

In addition, listing the files explicitly (and thus being able to change the list order) is what allowed me to recently discover the suboptimal behaviour described in #1035 . 

So I would prefer not to use glob, but obviously, if that's a blocker for you, I'll change it.

File: core/lowering/passes/CMakeLists.txt (hunk 4 above)
Comment (gcuendet) [reply core/lowering/passes/CMakeLists.txt:2]: see [my previous answer](https://github.com/pytorch/TensorRT/pull/1058#discussion_r872059785)

**Comment by gcuendet** [comment]
> @gcuendet Can you sign your commits?

I thought I did. If you look at my commit messages, they are all signed. What is it that is missing? Is there anything else to do than just `-s` when committing?

**Comment by gcuendet** [comment]
> Hi @gcuendet!
> 
> Thank you for your pull request and welcome to our community.
> 
> # Action Required
> In order to merge **any pull request** (code, docs, etc.), we **require** contributors to sign our **Contributor License Agreement**, and we don't seem to have one on file for you.
> 
> # Process
> In order for us to review and merge your suggested changes, please sign at https://code.facebook.com/cla. **If you are contributing on behalf of someone else (eg your employer)**, the individual CLA may not be sufficient and your employer may need to sign the corporate CLA.
> 
> Once the CLA is signed, our tooling will perform checks and validations. Afterwards, the **pull request will be tagged** with `CLA signed`. The tagging process may take up to 1 hour after signing. Please give it that time before contacting us about it.
> 
> If you have received this in error or have any questions, please contact us at [cla@fb.com](mailto:cla@fb.com?subject=CLA%20for%20pytorch%2FTensorRT%20%231058). Thanks!

This has come as kind of a surprise. In the CONTRIBUTING.md, there is no mention of a Meta CLA. I had the CLA included there reviewed by the legal department of my company before considering to contribute, but of course, now that the CLA has changed, I'll need a new round of reviewing. This might take time 😞

**Comment by narendasan** [comment]
@gcuendet Yeah sorry about that, it recently changed (like late last week) as we are now working with Meta on co-developing this project so the requirements to upstream code now includes the Meta CLA instead of DCO (https://pytorch.org/blog/a-contributor-license-agreement-for-pytorch/).

File: docsrc/tutorials/installation.rst [hunk 6]
```diff
@@ -238,6 +238,34 @@ To build using the pre-CXX11 ABI use the ``pre_cxx11_abi`` config
...
+      libtorch with `-DTorch_DIR=<path to libtorch>/share/cmake/Torch`
+    * If TensorRT is not found with the provided cmake finder, specify `-DTensorRT_ROOT=<path to TensorRT>`
+    * Finally, configure and build the project in a build directory of your choice with the following command
+      from the root of Torch-TensorRT project:
+
+    .. code-block:: shell
+
+        cmake -S. -B<build directory> \
+            [-DCMAKE_MODULE_PATH=cmake/Module] \
+            [-DTorch_DID=<path to libtorch>/share/cmake/Torch] \
```
Comment (narendasan) [inline docsrc/tutorials/installation.rst:264]: Just for my own understanding, why is it called `Torch_DID`?

File: core/partitioning/shape_analysis.cpp [hunk 7]
```diff
@@ -1,168 +1,169 @@
...
-    torch::jit::ConstantPooling(seg_block.g());
-    getSegmentsOutputByRunning(seg_block, example_tensor_map, partition_info);
-  }
-  return;
-}
-
-} // namespace partitioning
-} // namespace core
-} // namespace torch_tensorrt
+#include "core/partitioning/shape_analysis.h"
```
Comment (narendasan) [inline core/partitioning/shape_analysis.cpp:1]: Any idea why this file is changed?

File: core/conversion/converters/CMakeLists.txt (hunk 3 above)
Comment (narendasan) [reply core/conversion/converters/CMakeLists.txt:7]: Its fine by me as long as there is a reason, just means we will need to make sure to update both bazel and cmake as we add files

**Review by narendasan — COMMENTED** [review]
We probably want to export the internal APIs to register custom converters as well (see https://github.com/pytorch/TensorRT/blob/9e5fedf117d33fd3564e5c920ea2b058a994b127/examples/custom_converters/elu_converter/csrc/elu_converter.cpp#L6 for how it is used)

File: core/partitioning/shape_analysis.cpp (hunk 7 above)
Comment (gcuendet) [reply core/partitioning/shape_analysis.cpp:1]: Yes. I think the line endings across files in the repo are inconsistent. Apparently the line endings of that file changed when I committed.
Do you enforce a particular line endings? Which one should it be?

Comment (narendasan) [reply core/partitioning/shape_analysis.cpp:1]: We use UNIX line endings (LF)

File: docsrc/tutorials/installation.rst [hunk 8]
```diff
@@ -238,6 +238,34 @@ To build using the pre-CXX11 ABI use the ``pre_cxx11_abi`` config
...
+
+    * Latest releases of CMake and instructions on how to install are available for different platforms 
+      [on their website](https://cmake.org/download/).
+
+A few useful CMake options include:
+
+    * CMake finders for TensorRT and cuDNN are provided in `cmake/Modules`. In order for CMake to use them, pass
+      `-DCMAKE_MODULE_PATH=cmake/Modules` when configuring the project with CMake.
+    * Libtorch provides its own CMake finder. In case CMake doesn't find it, pass the path to your install of 
+      libtorch with `-DTorch_DIR=<path to libtorch>/share/cmake/Torch`
```
Comment (narendasan) [inline docsrc/tutorials/installation.rst:255]: Does this build system create something similar for TorchTRT? It would be nice if we could include this in the builds like we do for bazel.

Comment (gcuendet) [reply docsrc/tutorials/installation.rst:255]: Yes, it does. That's done [here](https://github.com/gcuendet/Torch-TensorRT/blob/700831fe15a09b6a3c4048bb6a79280667e03300/CMakeLists.txt#L31). It basically generates a `torchtrtConfig.cmake` that allows a consumer to do 
```
find_package(torchtrt)
``` 
in his CMakeLists.txt and then use the Cmake targets directly.

**Comment by gcuendet** [comment]
> 



> We probably want to export the internal APIs to register custom converters as well (see

What do you mean exactly with that?
Do you mean that there should be a (public) cmake target that allows a consumer to link against internal libraries and include their headers (typically on linux)? Or do you mean that the symbols should be exported (on windows)? 
Is there any plan to refactor that part and maybe better expose that functionality in the public API?

**Comment by narendasan** [comment]
> > 
> 
> > We probably want to export the internal APIs to register custom converters as well (see
> 
> What do you mean exactly with that? Do you mean that there should be a (public) cmake target that allows a consumer to link against internal libraries and include their headers (typically on linux)? Or do you mean that the symbols should be exported (on windows)? Is there any plan to refactor that part and maybe better expose that functionality in the public API?

I mean export the symbols. I don't think we have any plans to change this. We want to keep it separate from the user level API most people use.

File: cpp/include/torch_tensorrt/ptq.h [hunk 9]
```diff
@@ -314,7 +315,7 @@ class Int8CacheCalibrator : Algorithm {
  */
 
 template <typename Algorithm = nvinfer1::IInt8EntropyCalibrator2, typename DataLoader>
-TORCHTRT_API inline Int8Calibrator<Algorithm, DataLoader> make_int8_calibrator(
+inline Int8Calibrator<Algorithm, DataLoader> make_int8_calibrator(
```
Comment (narendasan) [inline cpp/include/torch_tensorrt/ptq.h:318]: Do these need to be exported as `TORCHTRT_API`? Seems like in #1091 they are not being exported properly

File: docsrc/tutorials/installation.rst [hunk 10]
```diff
@@ -238,6 +238,34 @@ To build using the pre-CXX11 ABI use the ``pre_cxx11_abi`` config
...
+    * If TensorRT is not found with the provided cmake finder, specify `-DTensorRT_ROOT=<path to TensorRT>`
+    * Finally, configure and build the project in a build directory of your choice with the following command
+      from the root of Torch-TensorRT project:
+
+    .. code-block:: shell
+
+        cmake -S. -B<build directory> \
```
Comment (andi4191) [inline docsrc/tutorials/installation.rst:262]: I am running into following error with the cmake command:

`-- Installing: /usr/local/lib/cmake/torchtrt/torchtrtTargets-noconfig.cmake
CMake Error at cpp/bin/torchtrtc/cmake_install.cmake:46 (file):
  file INSTALL cannot find
  "/Torch-TensorRT/build/cpp/bin/torchtrtc/torchtrtc": No such file or
  directory.
Call Stack (most recent call first):
  cpp/bin/cmake_install.cmake:47 (include)
  cpp/cmake_install.cmake:148 (include)
  cmake_install.cmake:52 (include)`
  
  There isn't any torchtrtc generated at the cpp/bin/torchtrtc/ path. I see it directly being generated at bin/

File: docsrc/tutorials/installation.rst [hunk 11]
```diff
@@ -238,6 +238,34 @@ To build using the pre-CXX11 ABI use the ``pre_cxx11_abi`` config
...
+            [-DCMAKE_MODULE_PATH=cmake/Module] \
+            [-DTorch_DIR=<path to libtorch>/share/cmake/Torch] \
```
Comment (andi4191) [inline docsrc/tutorials/installation.rst:264]: How will it work if the container already has a Torch library installed?

Can we take care of the suffix share/cmake/Torch internally?

File: docsrc/tutorials/installation.rst [hunk 12]
```diff
@@ -238,6 +238,34 @@ To build using the pre-CXX11 ABI use the ``pre_cxx11_abi`` config
...
+            [-DTensorRT_ROOT=<path to TensorRT>] \
+            [-DCMAKE_BUILD_TYPE=Debug|Release]
```
Comment (andi4191) [inline docsrc/tutorials/installation.rst:266]: Can we hardcode one by default internally in the CMakeLists.txt?

**Review by andi4191 — CHANGES_REQUESTED** [review]
Running into issues with linux.

File: cpp/include/torch_tensorrt/ptq.h (hunk 9 above)
Comment (gcuendet) [reply cpp/include/torch_tensorrt/ptq.h:318]: I _think_ these do not need to be exported, since the implementation is in the header, so it's not that the symbols need to be found in a dll... 
I'll try to investigate a bit more and comment on the issue #1091 directly.

File: docsrc/tutorials/installation.rst (hunk 11 above)
Comment (gcuendet) [reply docsrc/tutorials/installation.rst:264]: If the container already has Torch installed somewhere standard, there should be nothing to do and the `find_package(Torch)` is going to find it. This particular `-DTorch_DIR` is only required since the `TorchConfig.cmake` generated by libtorch is not installed somewhere standard.

File: docsrc/tutorials/installation.rst (hunk 10 above)
Comment (gcuendet) [reply docsrc/tutorials/installation.rst:262]: Yes, absolutely, sorry about that, my install rule was buggy. That should be fixed now.

File: cpp/include/torch_tensorrt/ptq.h (hunk 9 above)
Comment (gcuendet) [reply cpp/include/torch_tensorrt/ptq.h:318]: For the record, I just commented on that issue.

File: docsrc/tutorials/installation.rst (hunk 8 above)
Comment (gcuendet) [reply docsrc/tutorials/installation.rst:255]: This behavior is now demonstrated in the examples CMakeLists.txt. I was added in [this commit](https://github.com/pytorch/TensorRT/pull/1058/commits/61bfec8e75a1312c8fb5481245d453296e446386).

**Comment by narendasan** [comment]
@gcuendet We are currently testing this internally. Hopefully it should be wrapped up soon.

**Comment by gcuendet** [comment]
> @gcuendet We are currently testing this internally. Hopefully it should be wrapped up soon.

@narendasan how is the internal testing going? Is there anything I can help with? 
I tried to request access to your slack organisation (filled in the form), but so far without success. If there is anything that we should discuss directly, maybe we can find another way? 
Let me know.

**Comment by peri044** [comment]
I'm facing errors while running this 
```
cmake -S. -B<build directory> \
    [-DCMAKE_MODULE_PATH=cmake/Module] \
    [-DTorch_DIR=<path to libtorch>/share/cmake/Torch] \
    [-DTensorRT_ROOT=<path to TensorRT>] \
    [-DCMAKE_BUILD_TYPE=Debug|Release]
```
The exact command I'm using 
```
cmake -S . -B build/ -DTensorRT_LIBRARY=/home/dperi/Downloads/public_trt/TensorRT-8.2.4.2/lib -DTorch_DIR=/home/dperi/Downloads/TensorRT/bazel-TensorRT/external/libtorch_pre_cxx11_abi/share/cmake/Torch/  -DTensorRT_INCLUDE_DIR=/home/dperi/Downloads/public_trt/TensorRT-8.2.4.2/include/ -DcuDNN_LINK_LIBRARY=/home/dperi/Downloads/cudnn/cudnn-8.2.1/lib64/ -DcuDNN_INCLUDE_DIRS=/home/dperi/Downloads/cudnn/cudnn-8.2.1/include/
```
Errors: 
```
CMake Error at cpp/CMakeLists.txt:96 (add_library):
  Target "torchtrt_runtime" links to target "TensorRT::nvinfer_plugin" but
  the target was not found.  Perhaps a find_package() call is missing for an
  IMPORTED target, or an ALIAS target is missing?


CMake Error at cpp/CMakeLists.txt:124 (add_library):
  Target "torchtrt_plugins" links to target "TensorRT::nvinfer_plugin" but
  the target was not found.  Perhaps a find_package() call is missing for an
  IMPORTED target, or an ALIAS target is missing?


CMake Error at core/plugins/CMakeLists.txt:2 (add_library):
  Target "core_plugins" links to target "TensorRT::nvinfer_plugin" but the
  target was not found.  Perhaps a find_package() call is missing for an
  IMPORTED target, or an ALIAS target is missing?
```
@gcuendet  anything I'm missing here ?

**Comment by gcuendet** [comment]
> I'm facing errors while running this
> 
> ```
> cmake -S. -B<build directory> \
>     [-DCMAKE_MODULE_PATH=cmake/Module] \
>     [-DTorch_DIR=<path to libtorch>/share/cmake/Torch] \
>     [-DTensorRT_ROOT=<path to TensorRT>] \
>     [-DCMAKE_BUILD_TYPE=Debug|Release]
> ```
> 
> The exact command I'm using
> 
> ```
> cmake -S . -B build/ -DTensorRT_LIBRARY=/home/dperi/Downloads/public_trt/TensorRT-8.2.4.2/lib -DTorch_DIR=/home/dperi/Downloads/TensorRT/bazel-TensorRT/external/libtorch_pre_cxx11_abi/share/cmake/Torch/  -DTensorRT_INCLUDE_DIR=/home/dperi/Downloads/public_trt/TensorRT-8.2.4.2/include/ -DcuDNN_LINK_LIBRARY=/home/dperi/Downloads/cudnn/cudnn-8.2.1/lib64/ -DcuDNN_INCLUDE_DIRS=/home/dperi/Downloads/cudnn/cudnn-8.2.1/include/
> ```
> 
> Errors:
> 
> ```
> CMake Error at cpp/CMakeLists.txt:96 (add_library):
>   Target "torchtrt_runtime" links to target "TensorRT::nvinfer_plugin" but
>   the target was not found.  Perhaps a find_package() call is missing for an
>   IMPORTED target, or an ALIAS target is missing?
> 
> 
> CMake Error at cpp/CMakeLists.txt:124 (add_library):
>   Target "torchtrt_plugins" links to target "TensorRT::nvinfer_plugin" but
>   the target was not found.  Perhaps a find_package() call is missing for an
>   IMPORTED target, or an ALIAS target is missing?
> 
> 
> CMake Error at core/plugins/CMakeLists.txt:2 (add_library):
>   Target "core_plugins" links to target "TensorRT::nvinfer_plugin" but the
>   target was not found.  Perhaps a find_package() call is missing for an
>   IMPORTED target, or an ALIAS target is missing?
> ```
> 
> @gcuendet anything I'm missing here ?

**TL;DR**
Did you try with the actual documented command? (i.e. setting `-DTensorRT_ROOT=<path to TensorRT>`)
```
cmake -S . -B build/ -DTensorRT_ROOT=/home/dperi/Downloads/public_trt/TensorRT-8.2.4.2/ -DTorch_DIR=/home/dperi/Downloads/TensorRT/bazel-TensorRT/external/libtorch_pre_cxx11_abi/share/cmake/Torch/ -DcuDNN_ROOT_DIR=/home/dperi/Downloads/cudnn/cudnn-8.2.1/
```

In more details:

You are getting the following error:
```
CMake Error at cpp/CMakeLists.txt:96 (add_library):
  Target "torchtrt_runtime" links to target "TensorRT::nvinfer_plugin" but
  the target was not found.  Perhaps a find_package() call is missing for an
  IMPORTED target, or an ALIAS target is missing?
```
As the message says, CMake cannot find a **target**, specifically one of TensorRT targets (`TensorRT::nvinfer_plugin`). Apparently, you tried to manually specify `-DTensorRT_LIBRARY` and `-DTensorRT_INCLUDE_DIR` but this PR relies on CMake **targets** rather than just paths to libraries and include directories (which is sometimes refer to as "modern" cmake).
Since TensorRT and cuDNN do not come with CMake finders by default (allowing to define these **targets**), I provide these finders in `cmake/Module`. If TensorRT and/or cuDNN are not installed in a default location (as in your case), the only thing these finders need to define the targets is the path to the root of the respective install locations, specified with `-DTensorRT_ROOT` and/or `-DcuDNN_ROOT_DIR`.

I hope this helps. Let me know if something is unclear.

**Comment by peri044** [comment]
Thanks @gcuendet. It works now. 
I tested on ubuntu and it works. Following things need to be documented in the CMake section of installation.rst so that the expectation is clear and these can be addressed in the future update
1) Looks like the python api cannot be built using cmake now (as naren mentioned `--use-cmake`)
2) CMake support for tests is missing. 
cc: @narendasan

**Review by peri044 — CHANGES_REQUESTED** [review]
Following things need to be documented in the CMake section of installation.rst so that the expectation is clear and these can be addressed in the future update

Looks like the python api cannot be built using cmake now (as naren mentioned --use-cmake)
CMake support for tests is missing.

**Comment by gcuendet** [comment]
> Following things need to be documented in the CMake section of installation.rst so that the expectation is clear and these can be addressed in the future update
> 
> Looks like the python api cannot be built using cmake now (as naren mentioned --use-cmake) CMake support for tests is missing.

I added one more line in the CMake section of the `installation.rst`, as requested.

Note that, as offered when I initially opened this PR, I can work on the python api compilation with CMake or the unit test support, but I will not do that as part of this PR. I first would like to see this one merged. 

Also, if you think there is a need for these additional things and would welcome my contributions for these (as opposed to doing them internally), I would like to be able to better communicate with you (typically by having access to your slack org).

**Comment by narendasan** [comment]
> I would like to be able to better communicate with you (typically by having access to your slack org).

@gcuendet We are on the PyTorch slack which you can join through this form https://docs.google.com/forms/d/e/1FAIpQLSeADnUNW36fjKjYzyHDOzEB_abKQE9b6gqqW9NXse6O0MWh0A/viewform
